### PR TITLE
Add raw 'counts' calibration to 'abi_l1b' reader

### PR DESCRIPTION
**Pull request by tommyjasmin**
Adds ability to return the raw counts as a calibration type. This is sometimes used by McIDAS clients, for example when copying images from an ADDE server.

 - [x] Tests added 
 - [x] Fully documented 
 - [X] Add your name to `AUTHORS.md` if not there already

### CODE REVIEW

**Comment by djhoese** [comment]
Failures are caused by a new version of xarray and bad usage of creating Dataset objects in one of Satpy's readers. Fixes are in https://github.com/pytroll/satpy/pull/1777 and will be merged. I don't see any failures of your tests so far.

File: satpy/etc/readers/abi_l1b.yaml [hunk 1]
```diff
@@ -102,6 +102,9 @@ datasets:
       reflectance:
         standard_name: toa_bidirectional_reflectance
         units: "%"
+      counts:
+        standard_name: raw_counts
```
Comment (djhoese) [inline satpy/etc/readers/abi_l1b.yaml:106]: Where is the name "raw_counts" from? The existing readers that have counts use `standard_name: counts`.

File: satpy/readers/abi_l1b.py [hunk 2]
```diff
@@ -38,21 +38,27 @@ class NC_ABI_L1B(NC_ABI_BASE):
...
-        if key['calibration'] == 'reflectance':
-            logger.debug("Calibrating to reflectances")
-            res = self._vis_calibrate(radiances)
-        elif key['calibration'] == 'brightness_temperature':
-            logger.debug("Calibrating to brightness temperatures")
-            res = self._ir_calibrate(radiances)
-        elif key['calibration'] != 'radiance':
-            raise ValueError("Unknown calibration '{}'".format(key['calibration']))
+        # For raw cal, don't apply scale and offset, use true file counts
+        if (key['calibration'] == 'counts'):
```
Comment (djhoese) [inline satpy/readers/abi_l1b.py:42]:
```suggestion
        if key['calibration'] == 'counts':
```

File: satpy/readers/abi_l1b.py [hunk 3]
```diff
@@ -38,21 +38,27 @@ class NC_ABI_L1B(NC_ABI_BASE):
...
+        }
+
+        try:
+            res = cal_dictionary[key['calibration']](radiances)
+        except KeyError:
+            raise ValueError("Unknown calibration '{}'".format(key['calibration']))
 
         # convert to satpy standard units
-        if res.attrs['units'] == '1':
+        if (res.attrs['units'] == '1') and (key['calibration'] != 'counts'):
```
Comment (djhoese) [inline satpy/readers/abi_l1b.py:61]:
```suggestion
        if res.attrs['units'] == '1' and key['calibration'] != 'counts':
```

File: satpy/etc/readers/abi_l1b.yaml (hunk 1 above)
Comment (djhoese) [reply satpy/etc/readers/abi_l1b.yaml:106]: Ah the standard_name from the YAML isn't actually used as far as I can tell. The calibrate method sets its own information. This should probably be updated to match that or removed.

File: satpy/readers/abi_l1b.py [hunk 4]
```diff
@@ -38,21 +38,27 @@ class NC_ABI_L1B(NC_ABI_BASE):
...
-            logger.debug("Calibrating to reflectances")
-            res = self._vis_calibrate(radiances)
-        elif key['calibration'] == 'brightness_temperature':
-            logger.debug("Calibrating to brightness temperatures")
-            res = self._ir_calibrate(radiances)
-        elif key['calibration'] != 'radiance':
-            raise ValueError("Unknown calibration '{}'".format(key['calibration']))
+        # For raw cal, don't apply scale and offset, use true file counts
+        if (key['calibration'] == 'counts'):
+            radiances = self.nc['Rad']
```
Comment (djhoese) [inline satpy/readers/abi_l1b.py:43]: I'm not a huge fan of something being called radiances and meaning two different things depending on the calibration. Others may disagree with me but what about a shortcut here like:

```
if key['calibration'] == 'counts':
    return self._raw_calibrate(self.nc['Rad'])
```

Then you can remove the `'counts': self._raw_calibrate` entry in the dict below.

Open for discussion, just a feeling.

File: satpy/tests/reader_tests/test_abi_l1b.py [hunk 5]
```diff
@@ -269,6 +269,42 @@ def test_vis_calibrate(self):
...
+        super(Test_NC_ABI_L1B_raw_cal, self).setUp(rad=rad)
+
+    def test_raw_calibrate(self):
+        """Test RAW calibration."""
+        res = self.reader.get_dataset(
+            make_dataid(name='C05', calibration='counts'), {})
+
+        # We expect the raw data to be unchanged
+        expected = res.data
+        self.assertTrue(np.allclose(res.data, expected, equal_nan=True))
```
Comment (djhoese) [inline satpy/tests/reader_tests/test_abi_l1b.py:298]: Could you add an assertion here that `res.data.dtype == np.int16`? Basically to make sure the data isn't cast to another data type. I think `==` is the preferred way of comparing dtypes now. Or maybe it should be `res.data.dtype.type`.

**Review by djhoese — CHANGES_REQUESTED** [review]

File: satpy/etc/readers/abi_l1b.yaml (hunk 1 above)
Comment (tommyjasmin) [reply satpy/etc/readers/abi_l1b.yaml]: Agreed, we made this change Dave.

File: satpy/readers/abi_l1b.py (hunk 2 above)
Comment (tommyjasmin) [reply satpy/readers/abi_l1b.py]: Agreed, change has been made Dave.

File: satpy/readers/abi_l1b.py (hunk 3 above)
Comment (tommyjasmin) [reply satpy/readers/abi_l1b.py]: Agreed, change has been made Dave.

File: satpy/tests/reader_tests/test_abi_l1b.py [hunk 6]
```diff
@@ -269,6 +269,43 @@ def test_vis_calibrate(self):
...
+        self.assertTrue(res.data.dtype == np.int16, "int16 data type expected")
```
Comment (djhoese) [inline satpy/tests/reader_tests/test_abi_l1b.py:299]:
```suggestion
        self.assertEqual(res.data.dtype, np.int16, "int16 data type expected")
```

**Review by djhoese — APPROVED** [review]
I made one small change to the test assert that you added. Once the tests pass I think we can merge this!

**Pull request merged**